ros_packages/voice_assistant/voice_assistant/audio_recorder.py

```diff
@@ -0,0 +1,191 @@
+OUTPUT_FILE_PATH = VOICE_ASSISTANT_DIRECTORY + "/audiofiles/output.wav"
+
+
+
+class AudioRecorderNode(Node):
+
+    def __init__(self):
+
+        super().__init__('audio_recorder')
+
+        self.goal_queue: deque[ServerGoalHandle] = deque()
+        self.goal_queue_lock = Lock()
+
+        # server for recording audio and transcribing it to text, via the public-api
+        self.record_audio_server = ActionServer(
+            self, 
+            RecordAudio, 
+            'record_audio', 
+            execute_callback = self.record_audio,
+            handle_accepted_callback=self.handle_accepted_goal,
+            cancel_callback = (lambda _ : CancelResponse.ACCEPT))
+
+        self.get_logger().info('Now running AUDIO RECORDER')
+
+
+
+    def handle_accepted_goal(self, goal_handle: ServerGoalHandle) -> GoalResponse:
+        """place a goal into the queue and start execution if the queue was empty before"""
+        with self.goal_queue_lock:
+            if not self.goal_queue: goal_handle.execute()
+            self.goal_queue.appendleft(goal_handle)
+
+
+    
+    def is_silent(self, data_chunk) -> bool:
+        """Check whether a chunk of frmaes is below the minimum volume threshold"""
+        as_ints = np.frombuffer(data_chunk, dtype=np.int16)
+        return np.abs(as_ints).mean() < SILENCE_VOLUME_THRESHOLD
+    
+
+    def create_result(self, text: str) -> RecordAudio.Result:
+        """create an action-result and initilaize execution of the next queued goal"""
+
+        # create an result object and return
+        result = RecordAudio.Result()
+        result.transcribed_text = text
+
+        # remove the current goal from the queue, and start execution of next queued goal
+        # (if one is present)
+        with self.goal_queue_lock:
+            self.goal_queue.pop()
+            if self.goal_queue: self.goal_queue[-1].execute()
+
+        return result
+
+
+
+    def record_audio(self, goal_handle: ServerGoalHandle) -> None:
+        """
+        record audio and transcribe it to text. The result of the goal will be the transcribed text.
+        exactly one feedback object is returned. The object is empty and indicated, that recording
+        was stopped and transcription has begun
+        """
+
+        request: RecordAudio.Goal = goal_handle.request
+        
+        # these values indicate, after how many silent chunks, the recording is interrupted,
+        # the 'before' value is used, if speech has not started yet
+        # the 'after' value if used, after speech has already started
+        # generally, the 'before' value should be greater than the 'afer' value
+        # speech is considered to have started, once the first non-silent chunk is recognized
+        max_silent_chunks_before = request.max_silent_seconds_before * CHUNKS_PER_SECOND
+        max_silent_chunks_after = request.max_silent_seconds_after * CHUNKS_PER_SECOND
```

Comment:
right now, since the before value is hardcoded to 8 seconds, and for the after value, the "pause_threshold" of personalities is used (which only allows values up to 3 seconds), it is currently not possible for the before-value to be smaller than the after-value. This comment is meant as a hint, to indicate that the before-value is meant to give the user some time, to acually start speaking and should therefore be realitvely long.